server/src/test/java/org/opensearch/common/xcontent/BaseXContentTestCase.java

```diff
@@ -1027,6 +1027,25 @@ protected void doTestBigInteger(JsonGenerator generator, ByteArrayOutputStream o
         }
     }
 
+    protected void doTestInvalidSurrogatePair(JsonGenerator generator, ByteArrayOutputStream os) throws Exception {
```

Comment:
I would recommend putting this test in `libs/x-content` since that is where the change is. I think the test would look something like in `org/opensearch/common/xcontent/XContentParserTests.java`:

```java
public void testInvalidSurrogatePair() throws Exception {
    for (XContentType xContentType : XContentType.values()) {
        final String unmatchedSurrogate = "\uD800";
        try (XContentBuilder builder = XContentBuilder.builder(xContentType.xContent())) {
            builder.value(unmatchedSurrogate);

            try (XContentParser parser = createParser(xContentType.xContent(), BytesReference.bytes(builder))) {
                assertEquals(XContentParser.Token.VALUE_STRING, parser.nextToken());
                final String parsedUnmatchedSurrogate = parser.text();
                MatcherAssert.assertThat("Expected the unmatched surrogate to be preserved or to be replaced with the unicode replacement character (U+FFFD)",
                    parsedUnmatchedSurrogate,
                    anyOf(equalTo(unmatchedSurrogate), equalTo("\uFFFD")));
            }
        }
    }
}
```

But it also highlights the major problem here that some of the encodings replace the unmatched surrogate with U+FFFD and others round-trip it as-is.